File: Hot 100/Two Pointers/0042_TrappingRainWater.java
class Solution {
    public int trap(int[] height) {
        int ret = 0;
        int l = 0, r = height.length - 1;
        int lMax = 0, rMax = 0;
        while (l < r) {
            lMax = Math.max(height[l], lMax);
            rMax = Math.max(height[r], rMax);
            if (height[l] < height[r]) {
                ret += lMax - height[l];
                l++;
            } else {
                ret += rMax - height[r];
                r--;
            }
        }
        return ret;
    }
}

File: Top 100 Liked/Two Pointers/0042_trapping_rain_water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        l, r = 0, len(height) - 1
        l_max, r_max = height[l], height[r]
        ret = 0
        while l < r:
            l_max = max(height[l], l_max)
            r_max = max(height[r], r_max)
            if height[l] < height[r]:
                ret += l_max - height[l]
                l += 1
            else:
                ret += r_max - height[r]
                r -= 1
        return ret
